syntactic-paths/src/main/java/com/palantir/util/syntacticpath/Path.java

```diff
@@ -51,54 +45,49 @@
  */
 public final class Path implements Comparable<Path> {
 
-    public static final Path ROOT_PATH = new Path(ImmutableList.<String>of(), true, true);
+    public static final Path ROOT_PATH = new Path(new ArrayList<String>(), true, true);
     public static final char SEPARATOR_CHAR = '/';
     public static final String SEPARATOR = "/";
     public static final String BACKWARDS_PATH = "..";
 
-    private static final Splitter PATH_SPLITTER = Splitter.on(SEPARATOR_CHAR).omitEmptyStrings();
-    static final Joiner PATH_JOINER = Joiner.on(SEPARATOR_CHAR);
     private static final char[] ILLEGAL_CHARS = new char[] {0};
-    private static final List<String> ILLEGAL_SEGMENTS = ImmutableList.of(".");
-
+    private static final List<String> ILLEGAL_SEGMENTS = new ArrayList<String>() {
+        {
+            add(".");
+        }
+    };
     private final List<String> segments;
     private final int size;
     private final boolean isAbsolute;
     private final boolean isFolder;
-    private final Supplier<String> stringRepresentation;
-    private final Supplier<Path> normalizedPath;
+    private final String stringRepresentation;
+    private Path lazyInitializedNormalizedPath;
 
     private Path(final List<String> segments, final boolean isAbsolute, boolean isFolder) {
         this.segments = segments;
         this.size = segments.size();
         this.isAbsolute = isAbsolute;
         this.isFolder = isFolder;
-        stringRepresentation = Suppliers.memoize(new Supplier<String>() {
-            @Override
-            public String get() {
-                return toStringInternal();
-            }
-        });
-        normalizedPath = Suppliers.memoize(new Supplier<Path>() {
-            @Override
-            public Path get() {
-                return normalizeInternal();
-            }
-        });
+        this.stringRepresentation = toStringInternal();
     }
 
     Path(String input) {
         this(checkAndSplit(input), input.startsWith(SEPARATOR), input.endsWith(SEPARATOR));
     }
 
     private static List<String> checkAndSplit(String path) {
-        return checkSegments(ImmutableList.copyOf(PATH_SPLITTER.split(checkCharacters(path))));
+        return checkSegments(split(checkCharacters(path)));
     }
 
     private static String checkCharacters(String path) {
-        if (StringUtils.containsAny(path, ILLEGAL_CHARS)) {
-            throw new IllegalArgumentException("Path contains illegal characters: " + path);
+        for (char c : path.toCharArray()) {
+            for (char illegal : ILLEGAL_CHARS) {
+                if (illegal == c) {
+                    throw new IllegalArgumentException("Path contains illegal characters: " + path);
```

Comment:
now that we're checking one by one, let's make the error better: `Path /foo/./bar contains illegal character: '.'`